Goodbye Bower, hello Browserify

dmaloneycalu opened the pull request:
@vokal/web Here's the base for the v6 changes. There are a couple of TODOS that will need to be addressed before acceptance, but I wanted to get this out after the discussion yesterday.

Few notes that may not be obvious:
- Requires Node v4, update kiddos!
- Still need to add grunt-svgsprite (unless we want to move it to a per-project inclusion)
- Will require changing error routing on cloudfront to exclude "/build"
- There will be a parallel PR to dominatr


No rush here, please ask questions

jrit commented:
please add truecolors too, see changes on my fork from today

jrit commented on grunt/browserify.js:
```diff
@@ -0,0 +1,77 @@
+            browserifyOptions: {
+                debug: grunt.config( "envName" ) !== "prod",
+                noParse: [
+                    "/node_modules/**/*.js"
```
I think this might be ignored by default, but check on that, same on `ignore` below.

dmaloneycalu replied:
Will look into this further.

jrit commented on grunt/connect.js:
```diff
@@ -0,0 +1,38 @@
+var rewrite = require( "connect-modrewrite" );
+var serveStatic = require( "serve-static" );
+var rewriteMiddleware = rewrite( [
+    "^[^\\.]*$ /index.html [L]"
```
what does this do exactly?

dmaloneycalu replied:
Any request that's not specifically for a file (has a dot in the url), returns the index file

jrit commented on grunt/jshint.js:
```diff
@@ -2,30 +2,24 @@ module.exports = function ()
+            src: [
+                "source/modules/*/*.js",
+                "source/modules/*/*/*.js",
+                "!source/modules/*/tests/**/*.*"
```
Shouldn't tests pass jshint?

dmaloneycalu replied:
Up to us, these were excluded before as well.

jrit replied:
I'd include them, jshint and jscs should probably apply to all our JS.

dmaloneycalu replied:
Alrighty, I'll add them and I think this also misses the base index file.

jrit commented on grunt/protractor_coverage.js:
```diff
@@ -1,14 +1,17 @@
     },
     local: {
         options: {
+            keepAlive: true,
```
by default we are using keepAlive with bail-fast right?

dmaloneycalu replied:
We discussed including bail-fast locally, I've not included it yet in the dominatr changes though.

jrit commented on grunt/robots.js:
```diff
@@ -1,32 +1,25 @@
+            return;
         }
+
+        // TODO: can we determine if there is a sitemap file or not?
```
sure, use node and check if the file is there

dmaloneycalu replied:
Is this a step we should forcibly prevent if we don't have a sitemap?

jrit replied:
There should probably always be a robots.txt, but if there is no sitemap then it shouldn't write that line out. I don't know enough about sitemaps to say if it is better to just have a default sitemap.txt that points at the index file so we never have the no-content issue but that map is incomplete, or we shouldn't include sitemap.txt unless the content of it is relatively complete. To be fair, we should actually be maintaining this file for all the projects where it is relevant anyway.

dmaloneycalu replied:
@StoicJester do you want to chime in here at all?

StoicJester replied:
If we disallow crawling, it won't even look at the sitemap file. Otherwise, it's as @jrit says, and we should be tailoring a sitemap file with at least a line pointing to the index file.

However, if you say that there is a sitemap file, and the crawler can't find one, there isn't really a negative repercussion to that, it will still attempt to crawl as much of the site as it can. You can leave this code in. No need to overcomplicate things when there's no real drawback.

jrit commented on grunt/watch.js:
```diff
@@ -1,60 +1,36 @@
         options: { cwd: "<%= ngtemplates.build.cwd %>" },
         files: "<%= ngtemplates.build.src %>",
-        tasks: [ "envLocal", "ngtemplates", "replace" ]
+        // TODO: cut browserify out of this?
```
that won't work because browserify bundles the templates into the JS file

dmaloneycalu replied:
I got that much, but I'm certain it'd be possible to reference the template file in `index.html` then when deploying, join it with the browserified files. I figured it'd get complicated and just left this TODO here

jrit replied:
Yeah, don't do that :smile:

jrit commented on grunt/watch.js:
```diff
@@ -1,60 +1,36 @@
 module.exports = {
-    options:
-    {
-        atBegin: true,
```
why was `atBegin` removed?

dmaloneycalu replied:
Because `watch` was moved to the `npm start` task, which runs a build prior to running watch, making it redundant

Tathanen replied:
Only if we always use the `npm start` task, which per https://github.com/vokal/dominatr-grunt/pull/29, we don't.

dmaloneycalu replied:
Is it difficult to run `grunt build watch` instead? I thought the point to not running `npm start` on a project was because it ran the installs every time. npm/bower installs have been removed from `npm start` and now it only runs `grunt build connect:local watch`, eg: build, start server, watch

dmaloneycalu commented:
I also apparently missed re-adding karma testing, oops.

jrit commented:
So, on Karma, I totally like Karma where it is useful, but I'm not really into all these projects we have where it just runs one test that loads the index and then says everything is cool before it runs Protractor which has the "real" tests. That is kind of a waste of time, but it is also somewhat confusing for people to add and remove the scripts for Karma. Can you set it up so it is easy to run tests for either or both Karma and Protractor?

dmaloneycalu commented:
@jrit added truecolors support

jrit commented:
TC looks correct. This means I have to make the TC repos public, doing that now.

dmaloneycalu commented:
@jrit added grunt-svg-sprite, only changes from prior implementation is using less instead of css for the file output, and I added a watch task for it (and removed `/svg-sprite` from the copy task srcs)

jrit commented:
@dmaloneycalu any blockers on here right now?

dmaloneycalu commented:
@jrit I think the only blocker is not having re-added Karma testing, since for most of our repositories it was just spoofing of a test.

jrit commented:
I think Karma should be easily available, but in dominatr the default `npm test` command shouldn't run it since as you note, there are usually no real tests associated.

dmaloneycalu commented:
@vokal/web does anyone else have any questions/comments to make here before I tidy up the README and Karma?

dmaloneycalu commented:
Devs, there's a small issue I've found converting a project site over to this with file revisioning when using generated favicons and a `manifest.json` file for android-chrome. The strings in that file escape the url slashes, so they're ignored by `filerev_replace`.

Few options for this:
- Remove the escaping in `manifest.json`, hope the favicon works on android-chrome
- ignore favicons when revisioning by moving them to an ignored folder, either `/images/favicon` or `/favicon`
- write something custom for filerev_replace to check for escaped characters (no)

For the most part, we integrated file revisoning for caching benefit. If there was any other reason to use it, please feel free to speak up. Because we're deploying to versioned folders in cloudfront, I'm uncertain how much we really need it anymore.

@jrit @Tathanen @StoicJester @Deanzie @CMDadabo

Tathanen commented:
Well, favicons are notoriously cache-sticky, so they're something where this might become pretty useful.  I'd opt for the removal of the escaping, or at the very least investigate why it's necessary in the first place.

CMDadabo commented:
Is this PR currently blocked until the escaping issue above is resolved?

dmaloneycalu commented:
@CMDadabo It is not blocked by that issue, I'm writing the README now. Once that's completed and approved, this can be merged. The escaping issue is going to be ignored for now as it's not relevant for most users.

CMDadabo commented:
Sounds good. I was wondering where this was because I'm needing to move away from Bower for testing the Angular 2 stuff.

dmaloneycalu commented:
You can point at my latest commit in your package file if you want, `"dominatr-grunt": "github:dmaloneycalu/dominatr-grunt#acdc8f1"`.

dmaloneycalu commented:
Big README change, can [preview it here](https://github.com/dmaloneycalu/dominatr-grunt)

Tathanen commented:
I'd actually explain those last bits in a sentence or two instead of naming them as self-explanatory if you possess prior knowledge.

StoicJester commented on README.md:
```diff
@@ -1,22 +1,341 @@
+- #### `/modules/index.js`
+  This is the main `.js` file and should be used to require any dependencies. Anything that needs to be set globally should be attached here.
+
+  All module folders (except `mocks`, see below) should be `require`d. At the end of the file, the `ngtemplates` output must also be required.
```
`require`d => `required`

StoicJester replied:
should make use of `require`.

dmaloneycalu replied:
how's about "should be included with `require`."

StoicJester commented:
README looks good

dmaloneycalu commented:
Added an email notification task and a few issues to this repository after some testing with this PR. If none of the issues is pressing for this PR, we can move forward once we get a couple more LGTM.

jrit commented:
LGTM, think we should merge at this point

dmaloneycalu commented:
Will wait for @Tathanen's approval then merge

Tathanen commented:
So much in here now, we'll shake out any further refinements just by usin it more.

LGTM
